ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java

```diff
@@ -98,19 +99,40 @@ private void doCopyRetry(FileSystem sourceFs, List<ReplChangeManager.FileInfo> s
     while (!pathList.isEmpty() && (repeat < MAX_COPY_RETRY)) {
       LOG.info("Attempt: " + (repeat+1) + ". Copying files: " + pathList);
       try {
+        // if its retrying after copy error, first regenerate the path list.
+        if (isCopyError) {
+          pathList = getFilesToRetry(sourceFs, srcFileList, destinationFs, destination, true);
+        }
         isCopyError = false;
         doCopyOnce(sourceFs, pathList, destinationFs, destination, useRegularCopy);
+        pathList = getFilesToRetry(sourceFs, srcFileList, destinationFs, destination, false);
       } catch (IOException e) {
+        if (repeat >= (MAX_COPY_RETRY - 1)) {
+          //no need to wait in the last iteration
+          break;
+        }
         // If copy fails, fall through the retry logic
+        LOG.info("file operation failed with error : " + e.getMessage());
+
+        if (!(e instanceof FileNotFoundException)) {
+          LOG.info(" sleep for 100 milliseconds before retry " + (repeat+1));
+          try {
+            Thread.sleep(100);
+          } catch (InterruptedException timerEx) {
+            LOG.info("sleep interrupted", timerEx.getMessage());
+          }
+          // looks like some network outrage, reset the file system object and retry.
+          sourceFs = pathList.get(0).getFileSystem(hiveConf);
+          destinationFs = destination.getFileSystem(hiveConf);
+        }
         isCopyError = true;
```

Comment:
We set isCopyError=true even if copy was actually success and getFilesToRetry failed. This may have some behaviour difference.